### PR TITLE
NCHW FP16 SIMD kernel

### PR DESCRIPTION
**Pull request by SundarRajan28**
- Adding NCHW FP16 SIMD kernel for normalization and buffer copy
- Changing the process timer to the loop beginnning

### CODE REVIEW

File: rocAL/rocAL/source/pipeline/master_graph.cpp [hunk 1]
```diff
@@ -796,13 +796,78 @@ MasterGraph::copy_out_tensor(void *out_ptr, RocalTensorFormat format, float mult
                     else if(output_data_type == RocalTensorDataType::FP16) {
                         half *output_tensor_16 = static_cast<half *>(out_ptr);
                         auto channel_size = w * h;
```
Comment (sampath1117) [inline rocAL/rocAL/source/pipeline/master_graph.cpp:799]: Add the condition for c!=3 and move the code from Line 799 under else case similar to case of F32 condition
https://github.com/SundarRajan28/MIVisionX/blob/nchw_fp16_kernel/rocAL/rocAL/source/pipeline/master_graph.cpp#L724

File: rocAL/rocAL/source/pipeline/master_graph.cpp [hunk 2]
```diff
@@ -796,13 +796,78 @@ MasterGraph::copy_out_tensor(void *out_ptr, RocalTensorFormat format, float mult
                     else if(output_data_type == RocalTensorDataType::FP16) {
```
Comment (sampath1117) [inline rocAL/rocAL/source/pipeline/master_graph.cpp]: Move { in Line 796 to next line to maintain same style as of F32

File: rocAL/rocAL/source/pipeline/master_graph.cpp [hunk 3]
```diff
@@ -796,13 +796,78 @@ MasterGraph::copy_out_tensor(void *out_ptr, RocalTensorFormat format, float mult
...
                         half *output_tensor_16 = static_cast<half *>(out_ptr);
                         auto channel_size = w * h;
+    #if (ENABLE_SIMD && __AVX2__)
+                        half *B_buf = output_tensor_16 + dest_buf_offset;
```
Comment (sampath1117) [inline rocAL/rocAL/source/pipeline/master_graph.cpp]: Rename 
B_buf, G_buf, R_buf 
as
B_buf_16, G_buf_16, R_buf_16

File: rocAL/rocAL/source/pipeline/master_graph.cpp [hunk 4]
```diff
@@ -796,13 +796,78 @@ MasterGraph::copy_out_tensor(void *out_ptr, RocalTensorFormat format, float mult
...
+                            *B_buf++ = (half) (in_buffer[0] * multiplier0) + offset0;
+                            *G_buf++ = (half) (in_buffer[1] * multiplier1) + offset1;
+                            *R_buf++ = (half) (in_buffer[2] * multiplier2) + offset1;
+                        }
+    #else
                         for (unsigned channel_idx = 0; channel_idx < c; channel_idx++) {
                             for (unsigned i = 0; i < channel_size; i++)
                                 output_tensor_16[dest_buf_offset + channel_idx * channel_size + i] =
                                         offset[channel_idx] + multiplier[channel_idx] *
                                                               (reverse_channels ? (half) (in_buffer[(c*i+c-channel_idx-1)])
```
Comment (sampath1117) [inline rocAL/rocAL/source/pipeline/master_graph.cpp]: Add space before and after operators

File: rocAL/rocAL/source/pipeline/master_graph.cpp [hunk 5]
```diff
@@ -796,13 +796,78 @@ MasterGraph::copy_out_tensor(void *out_ptr, RocalTensorFormat format, float mult
...
+                        __m256 pmul1 = _mm256_set1_ps(multiplier1);
+                        __m256 pmul2 = _mm256_set1_ps(multiplier2);
+                        __m256 padd0 = _mm256_set1_ps(offset0);
+                        __m256 padd1 = _mm256_set1_ps(offset1);
+                        __m256 padd2 = _mm256_set1_ps(offset2);
+                        unsigned int alignedLength = (channel_size & ~7);    // multiple of 8
+                        unsigned int i = 0;
+
+                        __m256 fR, fG, fB;
+                        float dstPtrTempR_ps[8], dstPtrTempG_ps[8], dstPtrTempB_ps[8];
```
Comment (sampath1117) [inline rocAL/rocAL/source/pipeline/master_graph.cpp]: Rename 
dstPtrTempR_ps[8], dstPtrTempG_ps[8], dstPtrTempB_ps[8];
as 
tempR[8], tempG[8]. tempB[8]

File: rocAL/rocAL/source/pipeline/master_graph.cpp [hunk 6]
```diff
@@ -796,13 +796,78 @@ MasterGraph::copy_out_tensor(void *out_ptr, RocalTensorFormat format, float mult
...
+                            }
+                            B_buf += 8;
+                            G_buf += 8;
+                            R_buf += 8;
+                            in_buffer += 24;
+                        }
+                        for (; i < channel_size; i++, in_buffer += 3) {
+                            *B_buf++ = (half) (in_buffer[0] * multiplier0) + offset0;
+                            *G_buf++ = (half) (in_buffer[1] * multiplier1) + offset1;
+                            *R_buf++ = (half) (in_buffer[2] * multiplier2) + offset1;
```
Comment (sampath1117) [inline rocAL/rocAL/source/pipeline/master_graph.cpp]: Change offset1 to offset2

File: rocAL/rocAL/source/pipeline/master_graph.cpp (hunk 1 above)
Comment (SundarRajan28) [reply rocAL/rocAL/source/pipeline/master_graph.cpp:799]: Resolved in latest commit

File: rocAL/rocAL/source/pipeline/master_graph.cpp (hunk 2 above)
Comment (SundarRajan28) [reply rocAL/rocAL/source/pipeline/master_graph.cpp]: Resolved in latest commit

File: rocAL/rocAL/source/pipeline/master_graph.cpp (hunk 3 above)
Comment (SundarRajan28) [reply rocAL/rocAL/source/pipeline/master_graph.cpp]: Resolved in latest commit

File: rocAL/rocAL/source/pipeline/master_graph.cpp (hunk 4 above)
Comment (SundarRajan28) [reply rocAL/rocAL/source/pipeline/master_graph.cpp]: Resolved in latest commit

File: rocAL/rocAL/source/pipeline/master_graph.cpp (hunk 5 above)
Comment (SundarRajan28) [reply rocAL/rocAL/source/pipeline/master_graph.cpp]: Resolved in latest commit

File: rocAL/rocAL/source/pipeline/master_graph.cpp (hunk 6 above)
Comment (SundarRajan28) [reply rocAL/rocAL/source/pipeline/master_graph.cpp]: Resolved in latest commit

File: rocAL/rocAL/source/pipeline/master_graph.cpp [hunk 7]
```diff
@@ -895,6 +968,7 @@ void MasterGraph::output_routine()
 #endif
     }
     try {
+        _process_time.start();
```
Comment (shobana-mcw) [inline rocAL/rocAL/source/pipeline/master_graph.cpp]: Remove this change. This is not part of this PR.

File: rocAL/rocAL/source/pipeline/master_graph.cpp [hunk 8]
```diff
@@ -793,15 +793,88 @@ MasterGraph::copy_out_tensor(void *out_ptr, RocalTensorFormat format, float mult
     #endif
                         }
                     }
-                    else if(output_data_type == RocalTensorDataType::FP16) {
+                    else if(output_data_type == RocalTensorDataType::FP16) 
+                    {
```
Comment (shobana-mcw) [inline rocAL/rocAL/source/pipeline/master_graph.cpp:797]: Align { wrt other code in this function.

File: rocAL/rocAL/source/pipeline/master_graph.cpp [hunk 9]
```diff
@@ -793,15 +793,88 @@ MasterGraph::copy_out_tensor(void *out_ptr, RocalTensorFormat format, float mult
...
                         half *output_tensor_16 = static_cast<half *>(out_ptr);
                         auto channel_size = w * h;
-                        for (unsigned channel_idx = 0; channel_idx < c; channel_idx++) {
-                            for (unsigned i = 0; i < channel_size; i++)
-                                output_tensor_16[dest_buf_offset + channel_idx * channel_size + i] =
-                                        offset[channel_idx] + multiplier[channel_idx] *
-                                                              (reverse_channels ? (half) (in_buffer[(c*i+c-channel_idx-1)])
-                                                                                : (half) (in_buffer[(c * i + channel_idx)]));
+                        if(c != 3)
+                        {
```
Comment (shobana-mcw) [inline rocAL/rocAL/source/pipeline/master_graph.cpp]: Move { to if condition.

**Review by shobana-mcw — COMMENTED** [review]
@sampath1117 Did you review the logic?
Is there any scope for optimization?

**Comment by sampath1117** [comment]
> @sampath1117 Did you review the logic? Is there any scope for optimization?

@shobana-mcw i reviewed the logic and it is proper
Regarding optimization i just realized we can use 
_mm256_fmadd_ps instead of _mm256_mul_ps  and _mm256_add_ps
https://github.com/SundarRajan28/MIVisionX/blob/nchw_fp16_kernel/rocAL/rocAL/source/pipeline/master_graph.cpp#L844
https://github.com/SundarRajan28/MIVisionX/blob/nchw_fp16_kernel/rocAL/rocAL/source/pipeline/master_graph.cpp#L847

@SundarRajan28 Please make this change

File: rocAL/rocAL/source/pipeline/master_graph.cpp (hunk 7 above)
Comment (SundarRajan28) [reply rocAL/rocAL/source/pipeline/master_graph.cpp]: Resolved in the latest commit

File: rocAL/rocAL/source/pipeline/master_graph.cpp (hunk 9 above)
Comment (SundarRajan28) [reply rocAL/rocAL/source/pipeline/master_graph.cpp]: Resolved in the latest commit

File: rocAL/rocAL/source/pipeline/master_graph.cpp (hunk 8 above)
Comment (SundarRajan28) [reply rocAL/rocAL/source/pipeline/master_graph.cpp:797]: Leaving it as it is

**Comment by SundarRajan28** [comment]
Added FMA SSE instructions for the kernel instead of using mul_ps and add_ps instructions
